src/librustdoc/json/conversions.rs

```diff
@@ -170,12 +171,24 @@ impl FromWithTcx<clean::TypeBindingKind> for TypeBindingKind {
     }
 }
 
-crate fn from_def_id(did: FakeDefId) -> Id {
+crate fn from_item_id(did: ItemId) -> Id {
+    struct DisplayDefId(DefId);
+
+    impl fmt::Display for DisplayDefId {
+        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+            write!(f, "{}:{}", self.0.krate.as_u32(), u32::from(self.0.index))
+        }
+    }
+
     match did {
-        FakeDefId::Real(did) => Id(format!("{}:{}", did.krate.as_u32(), u32::from(did.index))),
-        // We need to differentiate real and fake ids, because the indices might overlap for fake
-        // and real DefId's, which would cause two different Id's treated as they were the same.
-        FakeDefId::Fake(idx, krate) => Id(format!("F{}:{}", krate.as_u32(), u32::from(idx))),
+        ItemId::DefId(did) => Id(format!("{}", DisplayDefId(did))),
+        ItemId::Blanket { for_, trait_ } => {
+            Id(format!("b:{}-{}", DisplayDefId(trait_), DisplayDefId(for_)))
+        }
+        ItemId::Auto { for_, trait_ } => {
+            Id(format!("a:{}-{}", DisplayDefId(trait_), DisplayDefId(for_)))
+        }
+        ItemId::Primitive(krate) => Id(format!("p:{}", krate.as_u32())),
```

Comment:
This doesn't seem right, it will treat all primitives as the same.

Actually, won't this cause bugs in other parts of rustdoc? Why does this work for the HTML backend?

Reply:
I was wondering too haha :smile:

But yeah it might be a good idea to add the type of primitive here too. I only added the krate because it was the minimum required to make it work.

Reply:
Can you just wrap the impl ID? I'm not sure why this doesn't use DefId in the first place ...